crates/shared/src/price_estimation/native/coingecko.rs

```diff
@@ -111,46 +128,93 @@ impl CoinGecko {
             })?
             .0;
 
-        let price = prices
+        Ok(tokens
+            .iter()
+            .map(|token| {
+                (
+                    **token,
+                    prices
+                        .get(token)
+                        .ok_or(PriceEstimationError::NoLiquidity)
```

Comment:
Not sure the `NoLiquidity` case needs to be handled at this low level. There are multiple code paths that convert a missing token price lookup into such an error so I think you could basically just return the coingecko response here.

Reply:
Here we are converting the raw CoinGecko response into a domain  type `(H16,  NativePriceResult)`. What we do here is check if we got an answer for each token requested, and for those without an answer, we return `NoLiquidity`. CoinGecko works that way, if there is no answer for the token requested, it is because there is no liquidty.

We cannot pass the CoinGecko response directly, we need to standarize it to our internal domain type first.

Reply:
The point I'm trying to make is that you have to handle that `map.get(token)` returns `None` in any case. So I think it's should be perfectly fine to just have the convention that not finding the token is equivalent to finding it with the error `NoLiquidity`.
That way you could simply return the `Response` here without having to do any transformations.